airflow/operators/python.py

```diff
@@ -145,6 +148,142 @@ def execute_callable(self):
+                                   'it several times in a DAG, use the `copy` method.')
+
+        # If we have no DAG, reinitialize class to capture DAGContext and DAG default args.
+        if not self.has_dag():
+            self.__init__(python_callable=self.python_callable,
+                          multiple_outputs=self.multiple_outputs,
+                          **self._kwargs)
+
+        # Capture args/kwargs
+        self._op_args = args
+        self._op_kwargs = kwargs
+        self._called = True
+        return XComArg(self)
+
+    def copy(self, task_id: Optional[str] = None, **kwargs):
+        """
+        Create a copy of the task, allow to overwrite ctor kwargs if needed.
+
+        If alias is created a new DAGContext, apply defaults and set new DAG as the operator DAG.
+
+        :param task_id: Task id for the new operator
+        :type task_id: Optional[str]
+        """
+        if task_id:
+            self._kwargs['task_id'] = task_id
+        return _PythonFunctionalOperator(
+            python_callable=self.python_callable,
+            multiple_outputs=self.multiple_outputs,
+            **{**kwargs, **self._kwargs}
+        )
+
+    def execute(self, context: Dict):
+        return_value = self.python_callable(*self._op_args, **self._op_kwargs)
+        self.log.info("Done. Returned value was: %s", return_value)
+        if not self.multiple_outputs:
+            return return_value
+        if isinstance(return_value, dict):
+            for key, value in return_value.items():
+                self.xcom_push(context, str(key), value)
+        elif isinstance(return_value, (list, tuple)):
+            for key, value in enumerate(return_value):
+                self.xcom_push(context, str(key), value)
+        return return_value
+
+
+def task(python_callable: Optional[Callable] = None, **kwargs):
+    """
+    Python operator decorator. Wraps a function into an Airflow operator.
+    Accepts kwargs for operator kwarg. Will try to wrap operator into DAG at declaration or
+    on function invocation. Use alias to reuse function in the DAG.
+
+    :param python_callable: Function to decorate
+    :type python_callable: Optional[Callable]
+    :param multiple_outputs: if set, function return value will be
+        unrolled to multiple XCom values. List/Tuples will unroll to xcom values
+        with index as key. Dict will unroll to xcom values with keys as keys.
+        Defaults to False.
+    :type multiple_outputs: bool
+
+    """
+    def wrapper(f):
+        """Python wrapper to generate PythonFunctionalOperator out of simple python functions.
+        Used for Airflow functional interface
+        """
+        return _PythonFunctionalOperator(python_callable=f, task_id=f.__name__, **kwargs)
+    if callable(python_callable):
+        return wrapper(python_callable)
+    elif python_callable is not None:
+        raise AirflowException('No args allowed while using @task, use kwargs instead')
+    return wrapper
```

Comment:
```suggestion
    _PythonFunctionalOperator._validate_python_callable(python_callable)
    return wrapper(python_callable)
```

I think is what we should do -- let's validate this as _early_ as possible.

What use case/pattern does allowing it to be not specified give us?

Reply:
This will not work. If I pass arguments to `@task(multiple_outputs=True)` this will break.

Reply:
Is there any particular reason why we don't do simply:
```python
def task(*args, **kwargs):
    """
    Python operator decorator. Wraps a function into an Airflow operator.
    Accepts kwargs for operator kwarg. Will try to wrap operator into DAG at declaration or
    on function invocation. Use alias to reuse function in the DAG.
    """
    if args:
        raise AirflowException("No args allowed")
    
    def wrapper(f):
        """
        Python wrapper to generate PythonFunctionalOperator out of simple python functions.
        Used for Airflow functional interface
        """
        _PythonFunctionalOperator.validate_python_callable(f)

        @functools.wraps(f)
        def factory(*args, **f_kwargs):
            op = _PythonFunctionalOperator(
                python_callable=f, 
                task_id=f.__name__,
                op_args=args, 
                op_kwargs=f_kwargs, 
                **kwargs
             )
            return XComArg(op)
        return factory
    return wrapper
```

Reply:
Yes, this code will fail when you do:
```
@task 
def example(...)
  ...
``` 
decorators get function as the first arg. If we want to allow both kwargs and no kwargs we need to do this small and more verbose approach.

Reply:
What about?
```python
if len(args) > 1:
        raise AirflowException("No args allowed")
```
We do this in `apply_defaults`.

Reply:
I'm still not sure why do we want to make task signature more generic. Your proposal basically would mean that I can do: `@task('test')` I still need to check that the first arg (if exists) is a callable. I don't see much difference in declaring the wrapper function before or after the check.

Reply:
@casassg I'm a bit confused, can we not give a task_id as an argument and have that become the task_id? Ideally we wouldn't just want "task_1" "task_2" etc. if we can name them (unless I'm missing something which I probably am)

Reply:
@dimberman you can name them manually as well. `@task(task_id='my_task')` Added a test for this.

Reply:
Awesome thanks @casassg